### ISSUE TITLE
Textures drawn wrongly

### ISSUE THREAD

**Issue by Spikeone** [Return-To-The-Roots/s25client#787]
Nearly any texture is drawn wrongly at the moment (for example morass, mountain meadow)

![image](https://user-images.githubusercontent.com/1276342/34360470-7792f796-ea61-11e7-8422-08a745ddd19c.png)
(Left RttR, right S2Editor)

Maybe sizes or areas are loaded wrongly, reference: http://settlers2.net/

**Comment by Flamefire** [comment]
Can you create screen shots showing the problem for the same map? Can't see the problem

**Comment by Spikeone** [comment]
![image](https://user-images.githubusercontent.com/1276342/34361114-9fee1c7e-ea67-11e7-82d1-4f6354c2d7e2.png)

**Comment by Flamefire** [comment]
Might be fixed with the lua terrains. I think the faulty textures are those that are smaller for other landscapes. Hence they get stretched.
Till then I classify this as minor as I had to look **really** close to see the difference.

**Comment by Spikeone** [comment]
Isn't meant in an offensive way, but thats what seperates longtime players from normal players ;)

Although I agree, that I also haven't noticed the bug right at the beginning, I did quite fast after taking a look at the morass which looks blurry and somewhat less sharp. Still think with the lua update this should be fixed as well because if not, I fear that bug is kept forever :(

**Comment by Flamefire** [comment]
I actually want to use shaders for terrain drawing. We did not use them at the beginning as it would mean no support for old graphic cards but now this should not be a problem anymore. With shaders we can texture the whole world. No more cutting deformed triangles out of the texture, we simply use the whole texture tiled over the whole world. Then it doesn't matter what size they are, they will always look like the image used. Only problem: BB didn't decide what to do and did both: Triangles for the water/lava and tiles for everything else. IMO we should simply rotate the water/lava by 45° and use them as tiles too.

But one step first: Lua terrain will fix many terrain issues although we have to compromise: The priority approach for edges means, we cannot be 100% identical to the original. But IMO we should focus on a good game instead of pixel identity.

**Comment by Spikeone** [comment]
I could understand that if we didn't have a nearly original version. Currently I do not understand why that needs to be changed then as it worked (maybe not perfectly, I understand that) but for the original graphics one could just leave it working in my opinion. 
Widelands looks bad, they may use fast or better techniques but the game itself just looks bad. RttR always looked (nearly) as good as the original, just the shading was missing but thats okay. I also agree that 1:1 pixels isn't needed but changing more than one edge type, changing textures and resolutions of textures, flow directions - that just feels off to me. I also understand that for new player or players who do not know Settlers 2 (or the last time they played it was 10 years ago) do not care about such details, but I do and that was always the point of this project. For a Settlers alike game there has been widelands, for nearly a 1:1 copy there was RttR. 

And Settlers 2 had beautyful pixel graphics, thats an important part of the game, by removing that the game loses something which would be quite sad.

**Comment by Flamefire** [comment]
Reason why it has to be changed is maintenance. Till recently only greenland was working properly. Now all 3 are drawn kind of correct to the extent possible when putting all 3 landscapes together. If we want to make it better, we need different handling for the 3 landscape types. That's why I made the lua terrains, where we can define for each and every terrain individual data so they can be like they should (and are in the original)
However due to the number of combinations we have to compromise on things. E.g. edge priority: Use 1 number per terrain is better then creating a huge table with a combination of every terrain with every other terrain. Possible, but not maintainable or extensible.
Different texture sizes (from the original!!!) mean that some textures will be stretched or made smaller when using the current approach. I want to use the whole texture from S2 w/o stretching so it looks like the original. That is not a change but an improvement!

- Edge type: I don't want to actually change them. Just improve them by using the original graphics. It is literally only a few pixels changed (e.g. for savannah all that changes is a bit of sand under the original meadow edge so you don't see the hard cut)
- Changing textures: That was never intended. Again: I just want the full original texture to be used to avoid the deformation. We have no idea how BB did it back then, but probably through a lot of hard-coded branching. I actually spend a lot of time trying to figure out which part of the texture they use, but did no find it. So using the full texture they created seems like the best idea.
- Texture resolution: What? We are and will be using the originals. Only thing changed would be *how much* of it. Currently we cut out a part and use that which looks good enough in most cases. But e.g. the mountain stretching was always in RttR due to exactly that and we can't do nothing about it if we don't use more of the texture.

Bottom line: I'm working hard on staying as close to the original as reasonably possible. It's good to know where there are differences to decide if we can fix them with reasonable effort. I don't want to copy widelands in RttR, just see, how techniques used there could help us. We will always be S2.5!
